### PR TITLE
Implement API that delete artifacts from the cache

### PR DESCRIPTION
**Pull request by maggie-lou**
- Implement API to delete artifacts from the cache
- Add unit tests

Fixes buildbuddy-io/buildbuddy-internal#1505

### CODE REVIEW

File: enterprise/server/api/api_server.go [hunk 1]
```diff
@@ -315,6 +331,56 @@ func (s *APIServer) GetFile(req *apipb.GetFileRequest, server apipb.ApiService_G
...
+			return nil, status.InvalidArgumentErrorf("Invalid URL. Does not match expected actioncache URI pattern: %s", err.Error())
+		}
+		cacheType = interfaces.ActionCacheType
+	} else if parsedURL.Scheme == "bytestream" {
+		parsedResourceName, err = digest.ParseDownloadResourceName(urlStr)
+		if err != nil {
+			return nil, status.InvalidArgumentErrorf("Invalid URL. Does not match expected CAS URI pattern: %s", err.Error())
+		}
+		cacheType = interfaces.CASCacheType
+	} else {
```
Comment (tylerwilliams) [inline enterprise/server/api/api_server.go:363]: do we need the "schemes" in these urls? if we can distinguish them just by the "/ac/" part that is in the action cache urls, that might be a cleaner API for users to use / learn. Also prevents us from having to worry about them accidentally setting the "host" or anything in that part of the URL.

Comment (maggie-lou) [reply enterprise/server/api/api_server.go]: Done

File: enterprise/server/backends/distributed/distributed.go [hunk 2]
```diff
@@ -443,7 +443,13 @@ func (c *Cache) remoteWriter(ctx context.Context, peer, handoffPeer string, isol
 	}
 	return c.cacheProxy.RemoteWriter(ctx, peer, handoffPeer, isolation, d)
 }
-
+func (c *Cache) remoteDelete(ctx context.Context, peer string, isolation *dcpb.Isolation, d *repb.Digest) error {
+	if !c.config.DisableLocalLookup && peer == c.config.ListenAddr {
+		// No prefix necessary -- it's already set on the local cache.
```
Comment (bduffany) [inline enterprise/server/backends/distributed/distributed.go:448]: what is "prefix" referring to here?

File: enterprise/server/backends/gcs_cache/gcs_cache.go [hunk 3]
```diff
@@ -299,6 +300,9 @@ func (g *GCSCache) Delete(ctx context.Context, d *repb.Digest) error {
 	timer.ObserveDelete(err)
 	// Note, if we decide to retry deletions in the future, be sure to
 	// add a new metric for retry count.
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		return status.NotFoundErrorf("Key %s not found in gcs_cache", d.GetHash())
```
Comment (bduffany) [inline enterprise/server/backends/gcs_cache/gcs_cache.go:304]: nit: return the full digest representation for debugging purposes, and also preserve the context from the original error:

```suggestion
		return status.NotFoundErrorf("digest %s/%d not found in gcs_cache: %s", d.GetHash(), d.GetSizeBytes(), err)
```

File: enterprise/server/backends/memcache/memcache.go [hunk 4]
```diff
@@ -256,7 +257,11 @@ func (c *Cache) Delete(ctx context.Context, d *repb.Digest) error {
 	if err != nil {
 		return err
 	}
-	return c.mc.Delete(k)
+	err = c.mc.Delete(k)
+	if errors.Is(err, memcache.ErrCacheMiss) {
+		return status.NotFoundErrorf("Key %s not found in memcache", d.GetHash())
```
Comment (bduffany) [inline enterprise/server/backends/memcache/memcache.go:262]:
```suggestion
		return status.NotFoundErrorf("digest %s/%d not found in memcache: %s", d.GetHash(), d.GetSizeBytes(), err)
```

**Review by vadimberezniker — CHANGES_REQUESTED** [review]
Can you split the cache changes and the external API changes into separate PRs?

Some people use Graphite to stack their changes. I think Zoey has had good luck with it I would suggest asking her about it.

File: enterprise/server/backends/distributed/distributed.go [hunk 5]
```diff
@@ -973,7 +979,18 @@ func (c *Cache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error
 }
 
 func (c *Cache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("Not yet implemented.")
+	ps := c.readPeers(d)
+	for peer := ps.GetNextPeer(); peer != ""; peer = ps.GetNextPeer() {
+		err := c.remoteDelete(ctx, peer, c.isolation, d)
+		if err != nil {
+			if status.IsNotFoundError(err) {
+				c.log.Warningf("File not found, could not be deleted: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply enterprise/server/backends/distributed/distributed.go:987]: My understanding was that readPeers(digest) should only return the nodes that should have the file? @vadimberezniker can you confirm?

File: enterprise/server/backends/distributed/distributed.go (hunk 2 above)
Comment (maggie-lou) [reply enterprise/server/backends/distributed/distributed.go:448]: remote instance name + cache type - added to comment (in the new split up CR)

File: enterprise/server/backends/gcs_cache/gcs_cache.go (hunk 3 above)
Comment (maggie-lou) [reply enterprise/server/backends/gcs_cache/gcs_cache.go:304]: Done (in new CR)

File: enterprise/server/backends/memcache/memcache.go (hunk 4 above)
Comment (maggie-lou) [reply enterprise/server/backends/memcache/memcache.go:262]: Done (in new CR)

File: enterprise/server/api/api_server.go [hunk 6]
```diff
@@ -315,6 +331,56 @@ func (s *APIServer) GetFile(req *apipb.GetFileRequest, server apipb.ApiService_G
 	})
 }
 
+func (s *APIServer) DeleteFile(ctx context.Context, req *apipb.DeleteFileRequest) (*apipb.DeleteFileResponse, error) {
+	ctx, _ = prefix.AttachUserPrefixToContext(ctx, s.env)
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply enterprise/server/api/api_server.go:335]: Done (in new CR)

**Comment by maggie-lou** [comment]
Split up the PR using graphite:
https://github.com/buildbuddy-io/buildbuddy/pull/2349
https://github.com/buildbuddy-io/buildbuddy/pull/2350
https://github.com/buildbuddy-io/buildbuddy/pull/2351

**Pull request closed without merging**